components/properties/src/provider/bidi_data.rs

```diff
@@ -0,0 +1,265 @@
+/// LE means first byte is 0..=7, second byte 8..=15, third byte is 16..=23
+///  0..=20  Code point return value for Bidi_Mirroring_Glyph value
+///    extracted with: mask = 0x1FFFFF <=> [bytes[0], bytes[1], bytes[2] & 0x1F]
+///  21..=21 Boolean for Bidi_Mirrored
+///    extracted with: bitshift right by 21 followed by mask = 0x1 <=> (bytes[2] >> 5) & 0x1
+///  22..=23 Enum discriminant value for Bidi_Paired_Bracket_Type
+///    extracted with: bitshift right by 22 followed by mask = 0x3 <=> (bytes[2] >> 6) & 0x3
+///                    <=> (bytes[2] >> 6) b/c we left fill with 0s on bitshift right for unsigned
+///                         numbers and a byte has 8 bits
+#[doc(hidden)]
+/// needed for datagen but not intended for users
+#[derive(Copy, Clone, Hash, PartialEq, Eq, Debug)]
+#[repr(packed)]
+pub struct MirroredPairedBracketDataULE([u8; 3]);
+
+// Safety (based on the safety checklist on the ULE trait):
+//  1. MirroredPairedBracketDataULE does not include any uninitialized or padding bytes
+//     (achieved by `#[repr(transparent)]` on a type that satisfies this invariant)
+//  2. MirroredPairedBracketDataULE is aligned to 1 byte.
+//     (achieved by `#[repr(transparent)]` on a type that satisfies this invariant)
+//  3. The impl of validate_byte_slice() returns an error if any byte is not valid.
+//  4. The impl of validate_byte_slice() returns an error if there are extra bytes.
+//  5. The other ULE methods use the default impl.
+//  6. MirroredPairedBracketDataULE byte equality is semantic equality because all bits
+//     are used, so no unused bits requires no extra work to zero out unused bits
+unsafe impl ULE for MirroredPairedBracketDataULE {
+    #[inline]
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), ZeroVecError> {
+        if bytes.len() % 3 != 0 {
+            return Err(ZeroVecError::InvalidLength {
+                ty: "MirroredPairedBracketDataULE",
+                len: bytes.len(),
+            });
+        }
+        // Validate the bytes
+        #[allow(clippy::indexing_slicing)] // Won't panic because the chunks are always 3 bytes long
+        for byte_triple in bytes.chunks_exact(3) {
+            // Bidi_Mirroring_Glyph validation
+            let byte0 = byte_triple[0];
+            let byte1 = byte_triple[1];
+            let byte2 = byte_triple[2];
+            let mut mirroring_glyph_code_point: u32 = (byte2 & 0x1F) as u32;
+            mirroring_glyph_code_point = (mirroring_glyph_code_point << 8) | (byte1 as u32);
+            mirroring_glyph_code_point = (mirroring_glyph_code_point << 8) | (byte0 as u32);
+            let _mirroring_glyph = char::from_u32(mirroring_glyph_code_point)
+                .ok_or(ZeroVecError::VarZeroVecFormatError)?;
+
+            // skip validating the Bidi_Mirrored boolean since it is always valid
+
+            // assert that Bidi_Paired_Bracket_Type cannot have a 4th value because it only
+            // has 3 values: Open, Close, None
```

Comment:
I'll strengthen my Thought to an Issue since there are several ways to construct a MirroredPairedBracketDataULE with a value of 3 for Bidi_Paired_Bracket_Type and we shouldn't impose more strict requirements on the deserialization of it.

All you need to do is remove this check. The only restriction on the ULE needs to be that the bottom 21 bits are a valid code point.

Reply:
Suggestion: Change the whole contents of this loop to

```rust
    let [byte0, byte1, byte2] = *<&[u8; 3]>::try_from(byte_triple).unwrap();
    let u = u32::from_le_bytes([byte0, byte1, byte2 & 0x1F, 0]);
    char::try_from(u).map_err(|_| ZeroVecError::parse::<Self>())?;
```

Reply:
Yeah, we should either make this an open integer type, or close the enum.

Note that the discriminant here does not have the full u8 range.

Reply:
@Manishearth This was the check that I was skipping (b/c the enum is open) that we discussed together and changed as a result. I think you said that Bidi_Paired_Bracket_Type will not add a property value beyond Open/Close/None, so we should have a stronger assertion accordingly, regardless of whether the Rust enum is open or not.

Of course, the only enumerated property guaranteed by the spec to be closed is General_Category. Everything else is theoretically subject to adding new property values.

I'm not terribly opinionated here because my intuition is weak, but I want to make sure that we can explain why we choose whichever decision we make here.

Reply:
Talking with @Manishearth, he reminded me that one of the ULE invariants that has to be enforced is that we cannot have a ULE byte serialization that corresponds to invalid data in its original form.

That hasn't been a problem for our property enums because they are all open enums and they all serialize to their own separate ULE type. Anything is valid in both the enum discriminant and the byte serialization.

Here, we're doing this bit packing for MirroredPairedBracketData's ULE type, and we've only allocated 2 bits to represent Bidi_Paired_Bracket_Type. So we're assuming that no new property values are defined (if they did, they would show up in a pair, probably, and we wouldn't be able to represent 3+2 = 5 property values in the 2 bits we've allocated in the bit packing). So we definitely aren't treating this as an open enum, so we have to treat it as a closed one.

I started down the path of the design using a closed version of the enum in MirroredPairedBracketData, and making it private, in to enforce the invariant while also avoiding having to write a custom impl of DataBake, but I can't tell from what I've done if I've misunderstood something in my new code or if it's possible to avoid writing a custom impl of DataBake.

Reply:
>  and making it private

we shouldn't make provider module types private

> in to enforce the invariant while also avoiding having to write a custom impl of DataBake, but I can't tell from what I've done if I've misunderstood something in my new code or if it's possible to avoid writing a custom impl of DataBake

What's wrong with slapping a `#[derive(Bake)]` onto it?

What you need here is your BPTClosed enum with Bake/Serialize/Deserialize custom derive impls on it, and `TryFrom<BPT>` (and `From` in the other direction)

Reply:
I don't think we need both the open enum and the closed enum. One should be sufficient.

I also don't understand why the open enum is violating a ULE invariant. I mean, if you made a `BMBT(45)`, it gets serialized as `0b01` and then comes back as `BMBT(1)`, which is a little weird but is it an invariant? It is the equality invariant at best, which could in principle be solved by a custom `PartialEq` on the open enum. Right?

All that said, I'm perfectly fine with a closed enum here. We should be able to use it throughout. We can add back the `(byte3 & 0xC0) != 0xC0` check and then we're good to go.

Reply:
Ok, and just so I understand, the reason we are okay to get rid of the open enum `BidiPairedBracketType` is because we don't have any use cases currently to support just the enum on its own (ex: no ECMA-402 requirement), but instead our only use case is via this space efficient Bidi properties struct?

I can get rid of `BidiPairedBracketType` and just have the `Checked-` version, as described. I'd like to keep the names unchanged so that if we ever need to faithfully represent Bidi_Paired_Bracket_Type property values as they are, we can revive the names without confusion.

Reply:
SGTM. Name doesn't matter too much since it is only a data provider type not directly exposed through the primary API